### ISSUE TITLE
Elements that have only text get lost

### ISSUE THREAD

**Issue by michael-ring** [yunkot/pxl#2]
XML Data like this:

```
  <name>SAM-E_DFP</name>
   <description>Atmel SAME70 Series Device Support</description>

ends up as:
  <name />
   <description />
```

when I save parsed xml to a new file.

Reason is that this case is not handled at all in the code when parsing the xml

I've created a patch that works at least for my test-data:

```
diff --git a/Source/PXL.XML.pas b/Source/PXL.XML.pas
index 1e7587e..9e0ae0f 100644
--- a/Source/PXL.XML.pas
+++ b/Source/PXL.XML.pas
@@ -47,6 +47,7 @@ type
   private
     {$IFDEF AUTOREFCOUNT}[weak]{$ENDIF} FParent: TXMLNode;
     FName: StdString;
+    FText: StdString;

     FieldItems: array of TFieldItem;
     FieldItemCount: Integer;
@@ -111,6 +112,7 @@ type

     property Parent: TXMLNode read FParent;
     property Name: StdString read FName;
+    property Text: StdString read FText;

     property FieldCount: Integer read GetFieldCount;
     property Field[const Index: Integer]: PFieldItem read GetField;
@@ -167,6 +169,7 @@ begin
   Increment_PXL_ClassInstances;

   FName := AName;
+  FText := '';
   FParent := AParent;
 end;

@@ -628,7 +631,10 @@ begin
     Result := Result + AddSpaces(Spacing) + '</' + FName + '>'#13#10;
   end
   else
-    Result := Result + ' />'#13#10;
+    if FText <> '' then
+      Result := Result + '>' + FText + '</'+FName+'>'#13#10
+    else
+      Result := Result + ' />'#13#10;
 end;

 function TXMLNode.GenerateSourceCode: StdString;
@@ -851,6 +859,7 @@ end;
 function ParseXMLNode(const Root: TXMLNode; const Text: StdString; var TextPos: Integer): TXMLNode;
 var
   NodeName: StdString;
+  OldTextPos : Integer;
 begin
   // Process node name.
   NodeName := ScanForName(Text, TextPos);
@@ -879,6 +888,8 @@ begin
     // Full end of node.
     if HasTextPortion(Text, TextPos, '/>') then
     begin
+      if OldTextPos <> 0 then
+        Result.FText := copy(Text,OldTextPos,TextPos-OldTextPos);
       Inc(TextPos, 2);
       Exit;
     end;
@@ -887,6 +898,9 @@ begin
     if Text[TextPos] = '>' then
     begin
       Inc(TextPos);
+      Result.FText := '';
+      // Save the Position of a possible pure Text in the Tag
+      OldTextPos := TextPos;
       Break;
     end;

@@ -894,6 +908,8 @@ begin
     if IsNameCharacter(Text[TextPos]) then
     begin
       ParseXMLField(Result, Text, TextPos);
+      // Make sure that Attributes do not set a TextPosition
+      OldTextPos := 0;
       Continue;
     end;

@@ -936,8 +952,10 @@ begin
     // End of node "</NODE>"
     if HasTextPortion(Text, TextPos, '</') then
     begin
+      // Fill Text Property when Element has no Children and does not consist of only Attributes
+      if (OldTextPos <> 0) then
+        Result.FText := copy(Text,OldTextPos,TextPos-OldTextPos);
       Inc(TextPos, 2);
-
       NodeName := ScanForName(Text, TextPos);
       if not SameText(NodeName, Result.Name) then
       begin
@@ -960,8 +978,9 @@ begin
     if Text[TextPos] = '<' then
     begin
       Inc(TextPos);
-
       ParseXMLNode(Result, Text, TextPos);
+      // Element has a Child node so it cannot have a valid text attribute
+      OldTextPos := 0;
       Continue;
     end;

```

**Comment by yunkot** [comment]
Actually, the main purpose of this XML parser was to work with config-like files, where only tags and fields are given, without text. Because of this, TXMLNode architecture cannot represent full logic of text among multiple tags. Something like:
```
<node>
  SomeText
    <node1>text1</node1>
  SomeText2
    <node2>text2</node2>
  SomeText3
</node>
```
The closest thing to parse aforementioned code would be to combine "SomeText", "SomeText2" and "SomeText3" into one text statement, but after saving, it would look like (in fact, there are multiple variants):
```
<node>
    <node1>text1</node1>
    <node2>text2</node2>
  SomeText
  SomeText2
  SomeText3
</node>
```
I know that this might not be a very common scenario, but it would be valid for something like HTML code.

**Comment by michael-ring** [comment]
You are right, this case will not work well, but other tools also show issues with this(tried xml tools of notepad++), so perhaps the value of beeing able to more completely parse a xml file outweights the relevancy of the corner cases.

**Comment by yunkot** [comment]
Actually, I think with minimal changes there is a way to properly parse multiple text and node chunks, but TXMLNode.Child in this case will be of a different type (e.g. node or text chunk). I'll be making the necessary changes and will update the trunk/master.

**Comment by michael-ring** [comment]
Great, thank you very much!

I was thinking about making my low level code for embedded microcontrollers wort togethwr wirh your lib, this could mean that if this works I would also be able to contribute code for some sensors, are you interested in contributions?

**Comment by yunkot** [comment]
Actually, I'm also interested on embedded micro-controllers target and in fact would definitely try porting the library there (btw, what CPU are you thinking about?); and, of course, any contributions are greatly welcome!

**Comment by michael-ring** [comment]
I am thinking about the Nucleo F401, F446 Boards from ST, the Freedom K64F and K22F Boards (This one is very similar to Teensy 3.1 and 3.2) and perhaps some day the Arduino Zero.
All those are ARM based Boards.

**Comment by yunkot** [comment]
That's very interesting, I would like to try participate in such experiments myself. This is getting off-topic, however, please let me know if there is any pending discussion on this elsewhere or how to make FPC work with one of these boards (and which ones are supported, to purchase one).

**Comment by michael-ring** [comment]
There is discussion from time to time on the lazarus / fpc-pascal / fpc-devel lists, some guy mentioned your lib there yesterday, that how I saw it.

The perfect start is perhaps to buy a Freedom K64F board as it is very easy to use and debug. The 2nd choice would be to use the Nucleo F401, debugging is a little less convinient with this board.

**Comment by yunkot** [comment]
This has been implemented in commit b1fcc97b1187b0a3c0a6e5dc2314eb64a29d6f4b.
Now each of TXMLNode.Child can be either TXMLNode or TXMLText. Please test this functionality and let me know if it works without any issues.

P.S. Regarding the other discussion, I'll try to get Freedom K64F for testing. If you'd like to continue this topic, my e-mail is [here](http://www.google.com/recaptcha/mailhide/d?k=01mYfnfc_ZbpA3NOt3YJxDog==&c=r99NTwsNkowLxr-neAN2qwxckBFJei80h7xQu8zw__U=).
